tool/compute_tag_analytics.py

```diff
@@ -1,196 +1,173 @@
-    col = firebase_client.collection(collection_root)
     time_start = time.perf_counter_ns()
-    for document_dict in documents:
-        push_document_to_firestore(firebase_client, collection_root, document_dict)
-    time_end = time.perf_counter_ns()
-    ms_elapsed = (time_end - time_start) / (1000 * 1000)
-    log.info (f"push_collection_to_firestore {collection_root} in {ms_elapsed} ms")
 
+    daily_metrics = {}
+    response_themes = coda_tags["response_themes"]
+    response_themes_ids = tag_ids(response_themes)
+    age_ids = tag_ids(coda_tags["age"])
+    gender_ids = tag_ids(coda_tags["gender"])
 
-def push_document_to_firestore(firebase_client, collection_root, data):
-    ref_path = data["__reference_path"]
-    doc_id = data["__id"]
-    sub_collections = data["__subcollections"]
+    for conversation in nook_conversations:
+        conversation_tags = conversation["tags"]
 
-    log.info (f"push_document_to_firestore {ref_path}")
+        if not ignore_stop and tag_to_tag_id("STOP") in conversation_tags:
+            continue
 
-    del data["__reference_path"]
-    del data["__id"]
-    del data["__subcollections"]
+        # get conversation demogs
+        age = "unknown"
+        age_tags = set(conversation_tags).intersection(age_ids)
+        if len(age_tags) == 1:
+            age = demogs.age_range(tag_id_to_name[list(age_tags)[0]])
+
+        gender = "unknown"
+        gender_tags = set(conversation_tags).intersection(gender_ids)
+        if len(gender_tags) == 1:
+            gender = tag_id_to_name[list(gender_tags)[0]]
+
+        # process response themes for each message
+        messages = conversation["messages"]
+        conversation_metrics = {}
+        for message in messages:
+            if message["direction"] == "MessageDirection.out":
+                continue
+            matched_tags = set(response_themes_ids).intersection(message["tags"])
+            if len(matched_tags) == 0:
+                continue
+            date = datetime.fromisoformat(message["datetime"]).date().isoformat()
+            date_metrics = conversation_metrics.get(date, {})
+
+            # demogs
+            if "age" not in date_metrics:
+                date_metrics["age"] = {}
+            date_metrics["age"][age] = True
+
+            if "gender" not in date_metrics:
+                date_metrics["gender"] = {}
+            date_metrics["gender"][gender] = True
+
+            # themes
+            if "response_themes" not in date_metrics:
+                date_metrics["response_themes"] = {}
+            for tag in matched_tags:
+                date_metrics["response_themes"][tag_id_to_name[tag]] = True
+            conversation_metrics[date] = date_metrics
+
+        # add conversation metrics to overall metrics
+        for date, date_metrics in conversation_metrics.items():
+            total_date_metrics = daily_metrics.get(date, empty_daily_metrics(response_themes))
+            for key1 in date_metrics.keys():
```

Comment:
This code is a bit surprising - I get what you're doing building something depending if it's a top level metric or a nested metric - but it might be worth a comment here?

Reply:
Thanks yeah, this is quite ugly code. I've added some more comments and also renamed the variables hoping that makes it a bit clearer.